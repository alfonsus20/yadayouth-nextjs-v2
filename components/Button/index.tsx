import cn from "classnames";
import React from "react";
import { CommonShape, CommonSize } from "../../types/theme";

type Appearance = "primary" | "secondary" | "tertiary";

type Props = {
  shape?: CommonShape;
  appearance?: Appearance;
  size?: CommonSize;
  children: React.ReactNode;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

const Button = ({
  shape = "rounded",
  appearance = "primary",
  size = "md",
  children,
  ...rest
}: Props) => {
  return (
    <button
      className={cn("hover:opacity-90 transition-opacity disabled:opacity-75", {
        rounded: shape === "rounded",
        "rounded-full": shape === "pill",
        "bg-blue": appearance === "primary",
        "px-10 py-2": size === "lg",
      })}
      {...rest}
    >
      {children}
    </button>
  );
};

export default Button;
