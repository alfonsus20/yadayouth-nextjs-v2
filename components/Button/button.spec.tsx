import Button from ".";
import { render, screen } from "@testing-library/react";
import { UserIcon } from "@heroicons/react/outline";

describe("Button unit test", () => {
  it("should match snapshot", () => {
    const container = render(<Button>Halo</Button>);
    expect(container).toMatchSnapshot();
  });

  it("should render correct shape", () => {
    const btnRounded = render(<Button shape="rounded">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain("rounded");
    btnRounded.unmount();

    const btnDefault = render(<Button>Halo</Button>);
    expect(screen.getByTestId("button").className).toContain("rounded");
    btnDefault.unmount();

    const btnPill = render(<Button shape="pill">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain("rounded-full");
    btnPill.unmount();
  });

  it("should render correct size", () => {
    const btnLarge = render(<Button size="lg">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain("px-10 text-lg");
    btnLarge.unmount();

    const btnMedium = render(<Button size="md">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain("px-6 text-md");
    btnMedium.unmount();

    const btnSmall = render(<Button size="sm">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain("px-3 text-sm");
    btnSmall.unmount();
  });

  it("should render correct appearance", () => {
    const btnDefault = render(<Button appearance="default">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-white text-gray-600"
    );
    btnDefault.unmount();

    const btnWithNoAppearanceProp = render(<Button>Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-blue text-white"
    );
    btnWithNoAppearanceProp.unmount();

    const btnPrimary = render(<Button appearance="primary">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-blue text-white"
    );
    btnPrimary.unmount();

    const btnSecondary = render(<Button appearance="secondary">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-yellow text-white"
    );
    btnSecondary.unmount();

    const btnTertiary = render(<Button appearance="tertiary">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-orange text-white"
    );
    btnTertiary.unmount();

    const btnDanger = render(<Button appearance="danger">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-red-light text-white"
    );
    btnDanger.unmount();

    const btnWarning = render(<Button appearance="warning">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-yellow-light text-white"
    );
    btnWarning.unmount();

    const btnInfo = render(<Button appearance="info">Halo</Button>);
    expect(screen.getByTestId("button").className).toContain(
      "bg-blue opacity-90 text-white"
    );
    btnInfo.unmount();
  });

  it("should be able to render left and right icon", () => {
    render(
      <Button
        leftIcon={<UserIcon className="w-5 h-5 text-black-light" />}
        rightIcon={<UserIcon className="w-5 h-5 text-black-light" />}
      >
        Halo
      </Button>
    );
    expect(screen.getByTestId("left-icon")).toBeVisible();
    expect(screen.getByTestId("right-icon")).toBeVisible();
  });
});
