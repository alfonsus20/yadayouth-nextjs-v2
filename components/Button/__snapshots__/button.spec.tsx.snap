// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button unit test should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="hover:opacity-90 transition-opacity disabled:opacity-75 disabled:cursor-not-allowed py-2 flex items-center rounded bg-blue text-white px-6 text-md"
        data-testid="button"
      >
        Halo
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="hover:opacity-90 transition-opacity disabled:opacity-75 disabled:cursor-not-allowed py-2 flex items-center rounded bg-blue text-white px-6 text-md"
      data-testid="button"
    >
      Halo
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
